File: Data Structures & Algorithms/largest-rectangle-in-histogram/submission-0.cpp
class Solution {
public:
    int largestRectangleArea(vector<int>& heights) {
        heights.push_back(0);
        int n = heights.size();
        stack<int> st;
        int res = 0;
        for (int i = 0; i < n; i++) {
            while (!st.empty() && heights[i] < heights[st.top()]) {
                int idx = st.top();
                st.pop();
                int height = heights[idx];
                int width = 0;
                if (st.empty()) width = i;
                else width = i - st.top() - 1;
                int area = height * width;
                if (area > res) {
                    res = area;
                }
            }
            st.push(i);
        }
        
        return res;
    }
};

File: Data Structures & Algorithms/max-water-container/submission-0.py
class Solution:
    def maxArea(self, heights: List[int]) -> int:
        i = 0
        j = len(heights) - 1
        maxArea = 0
        while i < j:
            curr = (j - i) * min(heights[i], heights[j])
            if maxArea < curr:
                maxArea = curr
            if heights[i] < heights[j]:
                i += 1
            else:
                j -= 1
        return maxArea